src/chain.h

```diff
@@ -147,6 +148,8 @@ class CBlockIndex
 
     //! (memory only) Sequential id assigned to distinguish order in which blocks are received.
     uint32_t nSequenceId;
+    //! (memory only) The time this node first heard about blocks
+    static std::map<const CBlockIndex*, int64_t> firstSeenTime;
```

Comment:
You may also want to either delete the copy constructor and assignment operator, or implement them. If not, the copies won't have the firstSeenTime information.